### PR TITLE
Leader-Board Implementation

### LINKED ISSUE #131: Implementing a leaderboard
## Description
As a User,
I would like to have a leaderboard which shows me my quantitative scores for various activities, and my rank in the leaderboard.

### Update [Required]
- [ ] Proper points implementation for each activity
- [ ] Credibility implementation (avoiding points increments on replaying the game)
- [ ] Ranking the users
- [ ] Making the UI accordingly.
- [ ] Logging the completion of the activity/game in the db (Backend integration)
- [ ] Unit testing the backend API with coverage > 80%
- [ ] Lint checks
- [ ] Travis check passes

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 2 mentors.

## Estimation
38-47 hours

### PR DESCRIPTION
**Pull request by lunayach**
### Description
Implemented the LeaderBoard component. Users are now ranked by their score in their respective badge categories. 

Fixes #131

### Type of Change:

- Code
- User Interface

- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)
- New feature (a non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Wrote new tests both in Mocha and Karma. Both existing and new tests pass.

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged 

- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

final8

### CODE REVIEW

File: routes/registration.js [hunk 1]
```diff
@@ -100,7 +101,10 @@ router.post('/', function(req, res) {
                         }).catch(error => {
                             res.status(500).json({error: errorCode.PCE006.message, code: errorCode.PCE006.code});
                         });
-                    });
+                        activityProgress.create({
+                            user_id: task.dataValues.user_id
```
Comment (Raxerz) [inline routes/registration.js]: What about default values for odd one out and drag and drop here?

File: src/app/introduction/activity-1/activity-1.component.ts [hunk 2]
```diff
@@ -69,8 +70,9 @@ export class HighlightActivityComponent implements OnInit {
                 });
                 this._sharedData.customSuccessAlert(this.alerts.activitySuccessMsg, this.alerts.activitySuccessTitle);
                 this.activityComplete = true;
-                if (!this.completed) { const currStage = 1;
-                  this._performanceService.openDialog(currStage); }
+          if (!this.completed) { const currStage = 1;
+                  this._performanceService.openDialog(currStage);
+                  this._leaderBoardService.updateLeaderBoard({activity: 'highlightDefinition', level: 'level1'})}
```
Comment (Raxerz) [inline src/app/introduction/activity-1/activity-1.component.ts]: Does the level attribute serve any purpose?

File: routes/registration.js (hunk 1 above)
Comment (lunayach) [reply routes/registration.js]: @Raxerz  Default values were set while declaring the model itself in ActivityProgress.js

File: src/app/header/header.component.html [hunk 3]
```diff
@@ -10,6 +10,7 @@
     </div>
   <div class="col-md-1">
     <img  [@newBadge]="state" [src]="badge?.image" [matTooltip]="badge?.name" (@newBadge.done) = "checkInitialState($event)" onError="this.src = 'assets/img/badges/default.gif'">
+    <i matTooltip="LeaderBoard" class="fa fa-user icons" (click)="loadLeaderBoard()"></i>
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/header/header.component.html]: @Raxerz  Okay! :+1:

File: src/app/introduction/activity-1/activity-1.component.ts (hunk 2 above)
Comment (lunayach) [reply src/app/introduction/activity-1/activity-1.component.ts]: @Raxerz  Ya, it serves a purpose. If you see **PointsConfig.ts** I have declared points to be given to activities as per levels. Currently, as all activities have one level, so we are giving *level1* flag. For multiple levels, we can have *level2*, *level3* flags respectively.

File: routes/registration.js (hunk 1 above)
Comment (Raxerz) [reply routes/registration.js]: Okay

File: src/app/introduction/activity-1/activity-1.component.ts (hunk 2 above)
Comment (Raxerz) [reply src/app/introduction/activity-1/activity-1.component.ts]: Is this being stored in the database?

File: config/passport.js [hunk 4]
```diff
@@ -51,32 +52,48 @@ module.exports = function(passport, models) {
...
                         })
                         .catch(function(err) {
                             return done(err);
                         });
-                })
+                    activityProgress.findOrCreate({where: {
+                        user_id: user.id
+                    }, defaults: {
+                        user_id: user.id,
+                        dragAndDrop: 0,
```
Comment (Raxerz) [inline config/passport.js]: Are the default values required here since it has been added as default in the models file

File: src/app/leader-board/leader-board.component.html [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    </thead>
+    <tbody *ngFor="let user of data; index as i">
+    <tr>
+      <td>{{i + 1}}</td>
+      <td>{{user.name}}</td>
+      <td>{{user.score}}</td>
+    </tr>
+    </tbody>
+  </table>
+</div>
```
Comment (Raxerz) [inline src/app/leader-board/leader-board.component.html]: new line

File: src/app/leader-board/leader-board.component.scss [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+#close {
+  text-align: right;
+  font-size: $font-size-large;
+  cursor: pointer;
+}
+
+#table {
+  text-align: center;
+}
```
Comment (Raxerz) [inline src/app/leader-board/leader-board.component.scss]: new line

File: src/app/leader-board/leader-board.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  });
+
+  it('should initiate the leaderBoard component successfully', () => {
+    expect(component).toBeTruthy();
+    expect(component.data).toEqual(mockData.leaderBoardData);
+    expect(component.image).toEqual(mockData.badge.image);
+    expect(component.badgeName).toEqual(mockData.badge.name);
+
+  });
+});
```
Comment (Raxerz) [inline src/app/leader-board/leader-board.component.spec.ts]: new line

File: src/app/leader-board/leader-board.component.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  ngOnInit() {
+    this.languageService.loadLanguage().subscribe(res => this.language = res.pcprepkit.common.leaderBoard.headings)
+  }
+
+  close() {
+    this.dialogRef.close();
+  }
+}
```
Comment (Raxerz) [inline src/app/leader-board/leader-board.component.ts]: new line

File: src/app/introduction/activity-1/activity-1.component.ts (hunk 2 above)
Comment (lunayach) [reply src/app/introduction/activity-1/activity-1.component.ts]: Currently, it's like this.
His score in a particular level is only going to be added when he completes that level for the first time. (not for activities which involve dynamic scores like Odd One Out etc.). So, passing the level flag from the game would ensure that his score is updated for that level.

**Comment by lunayach** [comment]
screenshot from 2018-07-10 00-17-17
screenshot from 2018-07-10 00-15-01
I have **new line** settings in both the files. For some reason, it's still not adding a new line on **GitHub**.

File: test/test.js [hunk 9]
```diff
@@ -258,7 +258,7 @@ describe('Testing APIs', function() {
                     .send(updatedStatus)
                     .expect(200)
                     .end(function(err, res) {
-                        res.body.info.should.equal('PCS003 : Successfully updated user progress status');
+                        res.body.info.should.equal('Successfully updated user progress status');
```
Comment (paavininanda) [inline test/test.js]: Similarly these changes can be taken up in another PR

File: src/app/introduction/activity-2/activity-2.component.ts [hunk 10]
```diff
@@ -15,17 +16,26 @@ export class PcpolicyComponent implements OnInit {
     language: any;
     email: String;
     private _status: object = {stage: 1, activity: 2};
-    public completed = true;
```
Comment (paavininanda) [inline src/app/introduction/activity-2/activity-2.component.ts]: How was this true till now..

Comment (lunayach) [reply src/app/introduction/activity-2/activity-2.component.ts]: @paavininanda this is Peace Corps Policy component. So, the user just sees it and the level is completed.

File: src/app/services/leaderBoard.service.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  getLeaderBoardData(badge: number): Observable<any> {
+    return this._apiService.post(this._getLeaderBoard, {badge: badge}).pipe(
+      map(response => response)
+    )
+  }
+
+  showLeaderBoard(data): void {
+    const dialogConfig = new MatDialogConfig();
+    dialogConfig.data = data;
+    dialogConfig.width = '700px';
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/services/leaderBoard.service.ts]: @paavininanda actually *width* here is *minimum width*. So for smaller screens, it adapts automatically.

File: src/app/leader-board/PointsConfig.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    level1: 100,
+    level2: 200,
+    level3: 300
+  },
+  stopBreed: {
+    level1: 100,
+    level2: 200
+  },
+  strideAgainst: {
+    level1: 150,
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/leader-board/PointsConfig.ts]: Okay. I thought we should vary this as per game difficulty. We may want to discuss this. @paavininanda

File: data/codes.json [hunk 13]
```diff
@@ -3,197 +3,220 @@
```
[the remark this answers is not in the document]
Comment (shreyans29) [reply data/codes.json]: yes.. the registration issue has been solved but we can create a new Issue for it.

File: src/app/introduction/activity-1/activity-1.component.ts [hunk 14]
```diff
@@ -69,8 +70,9 @@ export class HighlightActivityComponent implements OnInit {
                 });
                 this._sharedData.customSuccessAlert(this.alerts.activitySuccessMsg, this.alerts.activitySuccessTitle);
                 this.activityComplete = true;
-                if (!this.completed) { const currStage = 1;
-                  this._performanceService.openDialog(currStage); }
+          if (!this.completed) { const currStage = 1;
```
Comment (shreyans29) [inline src/app/introduction/activity-1/activity-1.component.ts]: const currStage = 1; new line..

File: src/app/introduction/activity-1/activity-1.component.ts (hunk 2 above)
Comment (shreyans29) [inline src/app/introduction/activity-1/activity-1.component.ts]: move the last brace to new line as well

File: src/app/leader-board/PointsConfig.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (shreyans29) [reply src/app/leader-board/PointsConfig.ts]: I think its tough to do that given different number of levels.

File: src/app/leader-board/leader-board.component.scss [hunk 15]
```diff
@@ -0,0 +1,21 @@
+@import "../styles/main";
+
+#body{
+  font-size: $font-size-med;
+  text-align: center;
+  text-transform: capitalize;
+}
+
+th{
```
Comment (shreyans29) [inline src/app/leader-board/leader-board.component.scss]: spacing

File: src/app/leader-board/leader-board.component.scss [hunk 16]
```diff
@@ -0,0 +1,21 @@
+@import "../styles/main";
+
+#body{
```
Comment (shreyans29) [inline src/app/leader-board/leader-board.component.scss]: spacing

File: src/app/services/leaderBoard.service.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (shreyans29) [reply src/app/services/leaderBoard.service.ts]: okay, can we change the variable name then ?

File: test/test.js (hunk 9 above)
Comment (shreyans29) [reply test/test.js]: yes..

File: src/app/header/header.component.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/header/header.component.html]: @shreyans29 didn't get you here. Can you please elaborate.

File: config/passport.js (hunk 4 above)
Comment (lunayach) [reply config/passport.js]: Yup! removed it.

File: data/codes.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (lunayach) [reply data/codes.json]: So, shall I move ahead and revert these changes?

File: src/app/services/leaderBoard.service.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/services/leaderBoard.service.ts]: Okay.

**Comment by lunayach** [comment]
Addressed the comments.

File: data/codes.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (shreyans29) [reply data/codes.json]: yes..

File: routes/api.js [hunk 17]
```diff
@@ -158,6 +159,151 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
...
+                    })
+
+            })
+            .catch(function() {
+                return res.status(500).json({error: errorCode.PCE032.message, code: errorCode.PCE032.code});
+            });
+    } else {
+        return res.status(400).json({error: errorCode.PCE033.message, code: errorCode.PCE033.code});
+    }
+});
```
Comment (shreyans29) [inline routes/api.js]: Are all  these covered by unit tests ?

File: src/app/malaria-101/activity-1/activity-1-2.component.ts [hunk 18]
```diff
@@ -108,10 +112,12 @@ export class MalariaLifeCycleComponent implements OnInit {
...
             this.activityComplete = true;
-          if (!this.completed) {const currStage = 3;
-            this._performanceService.openDialog(currStage); }
+          if (!this.completed) {
+            this._performanceService.openDialog(this.currStage);
+            this._leaderBoardService.updateLeaderBoard({activity: 'lifeCycle', level: 'level1'})
+          }
           this._sharedData.customSuccessAlert(this.alerts.activitySuccessMsg, this.alerts.activitySuccessTitle);
             this._dashboardService.updateProgressStatus(this._status).subscribe(response => {});
         } else if (arrLength === 6) {
```
Comment (shreyans29) [inline src/app/malaria-101/activity-1/activity-1-2.component.ts]: You need to add  6 and other constants as const.. It will improve code readability by lot.

File: src/app/malaria-101/activity-3/activity-3.component.ts [hunk 19]
```diff
@@ -23,8 +24,11 @@ export class OddOneOutComponent implements OnInit {
     private _subscription;
     private _data;
     private _status: object = {stage: 2, activity: 3};
+    private readonly pointsPerCorrectAnswer = 10;
```
Comment (shreyans29) [inline src/app/malaria-101/activity-3/activity-3.component.ts]: follow the CAPITALIZED_WITH_UNDERSCORES naming convention for all static and readonly constants/variable .

File: src/app/malaria-101/activity-3/activity-3.component.ts [hunk 20]
```diff
@@ -23,8 +24,11 @@ export class OddOneOutComponent implements OnInit {
...
+    private readonly QUESTION_NUMBER_FIVE = 5;
```
Comment (shreyans29) [inline src/app/malaria-101/activity-3/activity-3.component.ts]: this is the correct way, you have to make the change at other places.

File: routes/api.js (hunk 17 above)
Comment (lunayach) [reply routes/api.js]: Done.

**Comment by lunayach** [comment]
Addressed comments.

File: config/passport.js [hunk 21]
```diff
@@ -51,32 +52,46 @@ module.exports = function(passport, models) {
...
                 provider: 'google'
             }, defaults: {
-                    email: profile.emails[0].value,
-                    provider: 'google',
-                    google_token: accessToken,
-                    google_id: profile.id,
-                    fname: fname,
-                    lname: lname
-                }})
+                email: profile.emails[0].value,
```
Comment (Raxerz) [inline config/passport.js]: Why is the indentation changed?

File: config/passport.js [hunk 22]
```diff
@@ -51,32 +52,46 @@ module.exports = function(passport, models) {
...
                         return done(null, false, {info: errorCode.PCE010.message, code: errorCode.PCE010.code});
                     }
                     progress.findOrCreate({where: {
                         user_id: user.id
                     }, defaults: {
-                            user_id: user.id,
-                            stage: 0,
-                            activity: 0
-                        }})
+                        user_id: user.id,
```
Comment (Raxerz) [inline config/passport.js]: Same problem with the indentation.

File: database/models/activityProgress.js [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+        oddOneOut: {
+            type: Sequelize.INTEGER,
+            defaultValue: 0
+        },
+        user_id: {
+            type: Sequelize.INTEGER,
+            unique: true
+        }
+    }, {
+        tableName: 'activityProgress'
```
Comment (Raxerz) [inline database/models/activityProgress.js]: To maintain consistency, rename the table with underscore separation like 'activity_progress'

File: routes/api.js [hunk 24]
```diff
@@ -72,7 +73,7 @@ router.use(function(req, res, next) {
  * @param  {Object} (req, res)                                Anonymous function to handle request and response
  */
 router.get('/getUserInfo', authenticationHelpers.isAuthOrRedirect, (req, res) => {
-    res.status(200).json({user: req.user});
+    res.status(200).json({user: req.user.email});
```
Comment (Raxerz) [inline routes/api.js]: Why was this change made? This might affect the file like navbar.component.ts which consumes this API.

File: routes/api.js [hunk 25]
```diff
@@ -158,6 +159,151 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
     }
 });
 
+
+router.post('/leaderBoard', authenticationHelpers.isAuthOrRedirect, (req, res) => {
+    if (req.body) {
+        progress.findAll({
+            include: [localUser],
+            where: {badge: req.body.badge},
```
Comment (Raxerz) [inline routes/api.js]: Why is badge being used in where clause?
Are you ranking users based on a particular badge?

Comment (lunayach) [reply routes/api.js]: Exactly! @Raxerz, users would be ranked badge wise.

File: routes/api.js (hunk 24 above)
Comment (lunayach) [reply routes/api.js]: Actually, this API  was not used anywhere in the application until now. In navbar component, we use **get Username** API. 
From this API I get the user's email which I use in Leaderboard component to highlight the current user.

File: config/passport.js (hunk 21 above)
Comment (lunayach) [reply config/passport.js]: @Raxerz  The indentation was not correct. As locally, **eslint** was giving lint warnings with it. So, I changed that to address the warnings.

File: config/passport.js (hunk 22 above)
Comment (lunayach) [reply config/passport.js]: Same reason as above.

File: config/passport.js (hunk 21 above)
Comment (Raxerz) [reply config/passport.js]: Oh okay

File: src/app/leader-board/leader-board.component.scss [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+    font-size: $font-size-med;
+    text-align: center;
+    text-transform: capitalize;
+    color: $black;
+}
+
+th{
```
Comment (shreyans29) [inline src/app/leader-board/leader-board.component.scss]: spacing here too..

File: src/app/leader-board/leader-board.component.scss (hunk 16 above)
Comment (shreyans29) [reply src/app/leader-board/leader-board.component.scss]: this needs to be addressed. #body {

File: src/app/malaria-101/activity-3/activity-3.component.ts [hunk 27]
```diff
@@ -23,8 +24,11 @@ export class OddOneOutComponent implements OnInit {
     private _subscription;
     private _data;
     private _status: object = {stage: 2, activity: 3};
+    private readonly POINTS_PER_CORRECT_ANSWER = 10;
+    private readonly QUESTION_NUMBER_FIVE = 5;
```
Comment (Raxerz) [inline src/app/malaria-101/activity-3/activity-3.component.ts]: Isn't it better to change the variable name to LAST_QUESTION or something

File: src/app/services/leaderBoard.service.ts [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+
+@Injectable()
+
+export class LeaderBoardService {
+
+  private _baseAPIUrl = environment.baseURL + environment.apiEndpoint;
+  private _getLeaderBoard = this._baseAPIUrl + 'leaderBoard';
+  private _getScore = this._baseAPIUrl + 'user/score';
+  private _updateScore = this._baseAPIUrl + 'user/score/update';
+  private readonly dialogMinimumWidth = '700px';
```
Comment (Raxerz) [inline src/app/services/leaderBoard.service.ts]: can be a constant?

File: src/app/introduction/activity-1/activity-1.component.ts [hunk 29]
```diff
@@ -23,8 +24,9 @@ export class HighlightActivityComponent implements OnInit {
     public language: any;
     public alerts: any;
     public completed = false;
+    private readonly currStage = 1;
```
Comment (Raxerz) [inline src/app/introduction/activity-1/activity-1.component.ts]: Constant naming convention to be followed, please change it in the other files as well.

File: routes/api.js (hunk 25 above)
Comment (Raxerz) [reply routes/api.js]: @lunayach this is a good idea but was this implementation discussed with the team?
@Buddhiprabha @shreyans29 @paavininanda should this be badge based or for the entire set of activities/games?

File: src/app/introduction/activity-3/activity-3.component.ts [hunk 30]
```diff
@@ -368,7 +370,8 @@ export class PicturePuzzleComponent implements OnInit {
           const currStage = 2;
           this._performanceService.openDialog(currStage);
           this._badgeService.updateBadgeNumber(badgeNumber).subscribe(res => res);
-        }
+          this._leaderBoardService.updateLeaderBoard({activity: 'picturePuzzle', level: 'level1'})
```
Comment (shreyans29) [inline src/app/introduction/activity-3/activity-3.component.ts]: even here ?

File: src/app/malaria-101/activity-2/activity-2.component.ts [hunk 31]
```diff
@@ -16,6 +16,8 @@ export class DragdropComponent implements OnInit {
     public dropCheckDo: String;
     public dropCheckDont: String;
     language: any;
+    private score = 0;
+    private readonly pointsPerCorrectAnswer = 10;
```
[the remark this answers is not in the document]
Comment (shreyans29) [reply src/app/malaria-101/activity-2/activity-2.component.ts]: yes please change it to CAPS_WITH_UNDERSCORE

File: src/app/meds-n-labels/activity-1/activity-1.component.ts [hunk 32]
```diff
@@ -294,11 +296,13 @@ export class MatchmedsComponent implements OnInit {
...
-                              this._performanceService.openDialog(currStage); }
-                            this._dashboardService.updateProgressStatus(this._status).subscribe(response => {});
-                            this._infokitService.activateinfokit('match_meds').subscribe(res => {});
-                        }
+                          if (!this.completed) {
+                            this._performanceService.openDialog(this.currStage);
+                          }
+                              this._dashboardService.updateProgressStatus(this._status).subscribe(response => {});
+                              this._infokitService.activateinfokit('match_meds').subscribe(res => {});
+                              this._leaderBoardService.updateLeaderBoard({activity: 'matchMeds', level: 'level1'});
```
Comment (shreyans29) [inline src/app/meds-n-labels/activity-1/activity-1.component.ts]: same thing here..  no hardcoded strings.

File: src/app/meds-n-labels/activity-2/activity-2.component.ts [hunk 33]
```diff
@@ -168,9 +170,9 @@ export class MemoryGameComponent implements OnInit {
                 this.activityComplete = true;
                 if (!this.completed) {
                   const badgeNumber = 3;
```
Comment (shreyans29) [inline src/app/meds-n-labels/activity-2/activity-2.component.ts]: CAPS_WITH_UNDERSCORE naming convention

File: data/codes.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (lunayach) [reply data/codes.json]: Okay. They were reverted. :+1:

File: routes/api.js (hunk 25 above)
Comment (lunayach) [reply routes/api.js]: This was the idea I suggested in the proposal. This would ensure fair scoring between those who have played more stages than those who are still in earlier levels.

File: src/app/leader-board/leader-board.component.html [hunk 34]
```diff
@@ -0,0 +1,23 @@
...
+    <thead>
+    <tr>
+      <th>{{language?.rank}}</th>
+      <th>{{language?.name}}</th>
+      <th>{{language?.points}}</th>
+    </tr>
+    </thead>
+    <tbody *ngFor="let user of data; index as i">
+    <tr>
+      <td [ngClass] = "{ 'user' : user.userEmail === userEmail } ">{{i + 1}}</td>
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/leader-board/leader-board.component.html]: Here we are using the **user** from the **data** array in the For loop. It becomes convenient to check the condition of active user, while we are binding it in the template. 
It can be moved to the .ts file, then I check for this condition and associate some value to a some new variable, and then I again check equality of the variable, while binding it. 
So some condition needs to checked in the template in either case, as we want to control styling of the particular division.

File: src/app/malaria-101/activity-3/activity-3.component.ts [hunk 35]
```diff
@@ -23,8 +24,11 @@ export class OddOneOutComponent implements OnInit {
...
 
-    public activityComplete;
+
+  public activityComplete;
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/malaria-101/activity-3/activity-3.component.ts]: Done.

File: src/app/services/leaderBoard.service.ts (hunk 28 above)
Comment (lunayach) [reply src/app/services/leaderBoard.service.ts]: @Raxerz are you talking about **dialogMinimumWidth** here?

File: src/app/leader-board/leader-board.component.scss (hunk 26 above)
Comment (lunayach) [reply src/app/leader-board/leader-board.component.scss]: Done.

File: src/app/meds-n-labels/activity-2/activity-2.component.ts [hunk 36]
```diff
@@ -26,6 +27,7 @@ export class MemoryGameComponent implements OnInit {
     private _activity: number;
     private _stage: number;
     private _baseImgPath = '../../assets/img/Memory-game/';
+    private readonly currStage = 7;
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/meds-n-labels/activity-2/activity-2.component.ts]: done.

File: src/app/malaria-101/activity-3/activity-3.component.ts (hunk 27 above)
Comment (lunayach) [reply src/app/malaria-101/activity-3/activity-3.component.ts]: Done.

File: src/app/meds-n-labels/activity-2/activity-2.component.ts (hunk 36 above)
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/meds-n-labels/activity-2/activity-2.component.ts]: Done.

File: src/app/prevention/stop-the-breed/stop-the-breed.component.ts [hunk 37]
```diff
@@ -46,8 +48,11 @@ export class StopTheBreedComponent implements OnInit {
       this._dashboardService.updateProgressStatus({stage: 4, activity: 1})
         .subscribe(res => res);
       this.activityComplete = true;
-      if (!this.completed) { const currStage = 8;
-        this._performanceService.openDialog(currStage); }
+      if (!this.completed) {
+        const currStage = 8;
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/prevention/stop-the-breed/stop-the-breed.component.ts]: Done.

File: src/app/malaria-101/activity-2/activity-2.component.ts (hunk 31 above)
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/malaria-101/activity-2/activity-2.component.ts]: Done.

File: src/app/introduction/activity-1/activity-1.component.ts [hunk 38]
```diff
@@ -69,8 +71,10 @@ export class HighlightActivityComponent implements OnInit {
                 });
                 this._sharedData.customSuccessAlert(this.alerts.activitySuccessMsg, this.alerts.activitySuccessTitle);
                 this.activityComplete = true;
-                if (!this.completed) { const currStage = 1;
-                  this._performanceService.openDialog(currStage); }
+              if (!this.completed) {
+                this._performanceService.openDialog(this.currStage);
+                this._leaderBoardService.updateLeaderBoard({activity: 'highlightDefinition', level: 'level1'})
```
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/introduction/activity-1/activity-1.component.ts]: @shreyans29 I changed the hardcoded string of activities in all the files.
But, since **level** is dynamic and would depend on different levels, so this would become a global variable when we would have **muliple levels** PR merged.

File: src/app/meds-n-labels/activity-2/activity-2.component.ts (hunk 33 above)
Comment (lunayach) [reply src/app/meds-n-labels/activity-2/activity-2.component.ts]: Done.

**Comment by lunayach** [comment]
Addressed comments and LeaderBoard looks like this:
screenshot from 2018-07-22 02-33-04

File: routes/api.js (hunk 25 above)
Comment (Raxerz) [reply routes/api.js]: Okay

File: src/app/leader-board/leader-board.component.html (hunk 34 above)
[the remark this answers is not in the document]
Comment (Raxerz) [reply src/app/leader-board/leader-board.component.html]: Can't we have a function that takes it as a parameter in the ts file and returns the appropriate value?

File: src/app/services/leaderBoard.service.ts (hunk 28 above)
Comment (Raxerz) [reply src/app/services/leaderBoard.service.ts]: Yes. Its a constant right?

File: src/app/leader-board/leader-board.component.html (hunk 34 above)
[the remark this answers is not in the document]
Comment (shreyans29) [reply src/app/leader-board/leader-board.component.html]: I think what @Raxerz is suggesting will work and is a good way.

File: src/app/malaria-101/activity-2/activity-2.component.html [hunk 39]
```diff
@@ -8,7 +8,7 @@ <h2 class="sub-heading-1 text-center"> {{language?.heading}} </h2>
     <div class="content">
       <div class="row">
         <div class="col-xs-4 col-xs-offset-2">
-          <div dnd-droppable [className]="'panel panel-success ' + dropCheckDo" (onDropSuccess)="addTobox($event,box1)" (onDragOver)="checkDrop($event,box1)" (onDragLeave)="reset()">
+          <div dnd-droppable [className]="'panel panel-success '" (onDropSuccess)="addTobox($event,box1)"  (onDragLeave)="reset()">
```
[the remark this answers is not in the document]
Comment (shreyans29) [reply src/app/malaria-101/activity-2/activity-2.component.html]: got it..

File: src/app/services/leaderBoard.service.ts (hunk 28 above)
Comment (shreyans29) [reply src/app/services/leaderBoard.service.ts]: yep, its read only has to be. Naming convention..

Comment (lunayach) [reply src/app/services/leaderBoard.service.ts]: Okay! Done.

File: src/app/leader-board/leader-board.component.html (hunk 34 above)
[the remark this answers is not in the document]
Comment (lunayach) [reply src/app/leader-board/leader-board.component.html]: Okay! Got it. Done.

**Comment by lunayach** [comment]
Addressed comments.

File: routes/api.js (hunk 25 above)
Comment (Raxerz) [reply routes/api.js]: @lunayach what happens when there are no matching badges for any of the users?
How will the leaderboard be displayed? I tried this scenario and the leaderboard dialog was not displayed

Comment (lunayach) [reply routes/api.js]: @Raxerz  This scenario is not possible because all user must be having one badge. That is, even for a new user he is having a beginner badge.
The reason for not showing up of dialogue could be the columns of badge missing. You may try to reset database and try again.
